dockerversion/useragent.go

```diff
@@ -20,5 +24,50 @@ func DockerUserAgent() string {
 	httpVersion = append(httpVersion, useragent.VersionInfo{Name: "os", Version: runtime.GOOS})
 	httpVersion = append(httpVersion, useragent.VersionInfo{Name: "arch", Version: runtime.GOARCH})
 
-	return useragent.AppendVersions("", httpVersion...)
+	dockerUA := useragent.AppendVersions("", httpVersion...)
+	if len(upstreamUA) > 0 {
+		ret := insertUpstreamUserAgent(upstreamUA, dockerUA)
+		return ret
+	}
+	return dockerUA
+}
+
+// GetUserAgentFromContext returns the previously saved user-agent context stored in ctx, if one exists
+func GetUserAgentFromContext(ctx context.Context) string {
+	var upstreamUA string
+	if ctx != nil {
+		var ki interface{} = ctx.Value(httputils.UAStringKey)
+		if ki != nil {
+			upstreamUA = ctx.Value(httputils.UAStringKey).(string)
+		}
+	}
+	return upstreamUA
+}
+
+// escapeStr returns s with every rune in charsToEscape escaped by a backslash
+func escapeStr(s string, charsToEscape string) string {
+	var ret string
+	for _, currRune := range s {
+		appended := false
+		for _, escapeableRune := range charsToEscape {
+			if currRune == escapeableRune {
+				ret += "\\" + string(currRune)
```

Comment:
Why is this double escaped? If it encounters `;`, it should emit `\;`. If it encounters `\`, it should emit `\\`.